reps/2023-12-14-virtual-cluster.md

```diff
@@ -0,0 +1,452 @@
+  // One of PACK, SPREAD, or STRICT_SPREAD. These would be
+  // defined with respect to parent virtual nodes for nested
+  // clusters.
+  SchedulingPolicy scheduling_policy
+}
+```
+
+#### Job API
+
+Currently we have two ways to run a Ray job: ``ray.init()`` and ``ray job submit``. Both will take an optional parameter specifying the spec of the virtual cluster inside which the job will run. If unspecified, the default virtual cluster has zero min resources and infinite max resources meaning it can scale up to use the entire physical cluster resources.
+
+```python
+# Default virtual cluster
+# The job can use up to the entire physical cluster resources.
+ray.init()
+
+# The job can use at least 1 CPU and at most 8 CPUs.
+ray.init(virtual_cluster=VirtualCluster(flexible_resource_min={"CPU": 1}, flexible_resource_max={"CPU": 8}))
+
+# The job uses 1 A100 GPU.
+ray.init(virtual_cluster=VirtualCluster(fixed_size_nodes=[FixedSizeNode(resources={"GPU": 1}, parent_node_selector={"accelerator_type": In("A100")})]))
+```
+
+```shell
+# Default virtual cluster
+# The job can use up to the entire physical cluster resources.
+ray job submit -- python job.py
+
+# The job needs 2 * 1 GPU that are strict spreaded.
+ray job submit --virtual-cluster='{"fixed_size_nodes": [{"nodes": [{"resources": {"GPU": 1}}, {"resources": {"GPU": 1}}], "scheduling_policy": "STRICT_SPREAD"}]}' -- python job.py
+```
+
+Once a job is running inside a virtual cluster, it can use all the Ray APIs as if it's running inside its own Ray cluster.
+
+#### Placement Group API
+
+Since virtual clusters are nestable and support gang scheduling, they can be used to implement or replace placement groups.
+
+```python
+# Create a placement group with two bundles that are packed
+pg = placement_group([{"GPU": 4}, {"GPU": 1}], strategy="PACK")
+# Run the task inside the placement group
+task.options(scheduling_strategy=PlacementGroupSchedulingStrategy(placement_group=pg)).remote()
+
+# Create a virtual cluster with two virtual nodes that are packed
+vc = VirtualCluster(fixed_size_nodes=[FixedSizeNodes(nodes=[FixedSizeNode(resources={"GPU": 4}), FixedSizeNode(resources={"GPU": 1})], scheduling_policy=PACK)])
+# A nested virtual cluster inside the job virtual cluster
+with vc:
+  # Run the task inside the virtual cluster
+  task.remote()
+```
+
+```python
+# Create a placement group with two bundles that are strict spreaded
+pg = placement_group([{"GPU": 4}, {"GPU": 1}], strategy="STRICT_SPREAD")
+# Run the actor using the bundle 0 resources
+Actor.options(scheduling_strategy=PlacementGroupSchedulingStrategy(placement_group=pg, placement_group_bundle_index=0)).remote()
+
+# Create a virtual cluster with two virtual nodes that are strict spreaded
+vc = VirtualCluster(fixed_size_nodes=[FixedSizeNodes(nodes=[FixedSizeNode(resources={"GPU": 4}, labels={"bundle_index": "0"}), FixedSizeNode(resources={"GPU": 1}, labels={"bundle_index": "1"})], scheduling_policy=STRICT_SPREAD)])
```

Comment:
Nice, I like how we can get rid of the explicit bundle indexing in favor of labels.